coord/src/rmr/redise.c

```diff
@@ -5,38 +5,188 @@
  */
 
 #include "redise.h"
-#include "redise_parser/parse.h"
+
+typedef struct {
+  int startSlot;
+  int endSlot;
+  MRClusterNode node;
+} RLShard;
+
+static void MRTopology_AddRLShard(MRClusterTopology *t, RLShard *sh) {
+
+  for (int i = 0; i < t->numShards; i++) {
+    if (sh->startSlot == t->shards[i].startSlot && sh->endSlot == t->shards[i].endSlot) {
+      // New node in the same shard (slot range)
+      MRClusterShard_AddNode(&t->shards[i], &sh->node);
+      return;
+    }
+  }
+
+  // New shard
+  MRClusterShard csh = MR_NewClusterShard(sh->startSlot, sh->endSlot, 2);
+  MRClusterShard_AddNode(&csh, &sh->node);
+  MRClusterTopology_AddShard(t, &csh);
+}
+
+/* Error replying macros, in attempt to make the code itself readable */
+#define ERROR_FMT(fmt, ...)                          \
+  ({                                                 \
+    char *err;                                       \
+    __ignore__(rm_asprintf(&err, fmt, __VA_ARGS__)); \
+    RedisModule_ReplyWithError(ctx, err);            \
+    rm_free(err);                                    \
+  })
+
+#define ERROR_BADVAL(arg, val) ERROR_FMT("Bad value for %s: %s", arg, val)
+#define ERROR_EXPECTED(exp, arg) ERROR_FMT("Expected " exp " but got `%s`", arg)
+#define ERROR_MISSING(arg) RedisModule_ReplyWithError(ctx, "Missing value for " arg)
+
+#define ERROR_BAD_OR_MISSING(arg, ac_code)          \
+  ({                                                \
+    if (ac_code == AC_ERR_NOARG) {                  \
+      ERROR_MISSING(arg);                           \
+    } else {                                        \
+      ERROR_BADVAL(arg, AC_GetStringNC(&ac, NULL)); \
```

Comment:
Consider passing `ac` as argument to the macro so it will not break if we will change var names...